File: 1964-find-the-longest-valid-obstacle-course-at-each-position/1964-find-the-longest-valid-obstacle-course-at-each-position.cpp
class Solution {
public:
    vector<int> longestObstacleCourseAtEachPosition(vector<int>& obstacles) {
        int n = obstacles.size();
        vector<int> ans (n);
        vector<int> lis;
        for (int i = 0; i < n; i++) {
            int idx = upper_bound(lis.begin(), lis.end(), obstacles[i]) - lis.begin();
            if (idx == lis.size()) {
                lis.push_back(obstacles[i]);
                ans[i] = lis.size();
            } else {
                lis[idx] = obstacles[i];
                ans[i] = idx + 1;
            }
        }
        
        return ans;
    }
};

File: 1964-find-the-longest-valid-obstacle-course-at-each-position/1964-find-the-longest-valid-obstacle-course-at-each-position.java
class Solution {
    
    private int upperBound(List<Integer> lis, int val) {
        int lb = 0;
        int ub = lis.size();
        while(lb < ub) {
            int mid = lb + (ub - lb) / 2;
            if (val >= lis.get(mid)) {
                lb = mid + 1;
            } else {
                ub = mid;
            }
        }
        return ub;
    }
    
    public int[] longestObstacleCourseAtEachPosition(int[] obstacles) {
        int n = obstacles.length;
        int[] ans = new int[n];
        List<Integer> lis = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            int idx = upperBound(lis, obstacles[i]);
            if (idx == lis.size()) {
                lis.add(obstacles[i]);
                ans[i] = lis.size();
            } else {
                lis.set(idx, obstacles[i]);
                ans[i] = idx + 1;
            }
        }
        
        return ans;
    }
}